src/frequenz/sdk/timeseries/formulas/_ast.py

```diff
@@ -0,0 +1,205 @@
+# License: MIT
+# Copyright © 2025 Frequenz Energy-as-a-Service GmbH
+
+"""Formula AST nodes and evaluation logic."""
+
+from __future__ import annotations
+
+import abc
+import logging
+import math
+from collections.abc import AsyncIterator
+from dataclasses import dataclass
+from typing import Generic
+
+from typing_extensions import override
+
+from ..._internal._math import is_close_to_zero
+from .._base_types import QuantityT, Sample
+from ._functions import Function
+
+_logger = logging.getLogger(__name__)
+
+
+@dataclass
```

Comment:
I suggest making all these dataclasses `kw_only` for extra clarity. If not all, at least the ones taking 2 arguments of the same type, like `Add` (`left`, `right`).